File: LC/[1293] Shortest Path in a Grid with Obstacles Elimination.cpp
/*
* DP?: we can consider DP if we can only move down and right.  In this problem, user can go up and left (go back), this violates DP requirement (no afterward effects)
*/

class Solution {
     bool visited[41][41][1601];
public:
    int shortestPath(vector<vector<int>>& grid, int K) {
       int m=grid.size();
       int n=grid[0].size();
       if(m==1 && n==1) return 0;
       //Shortest distance -> BFS -> Queue
       queue<vector<int>>q;
       q.push({0,0,0});
       visited[0][0][0]=1;
       auto dir = vector<pair<int, int>>({{0,1},{0,-1},{1,0},{-1,0}});
       int step = 0;
       while(!q.empty()){
           int len = q.size();
           while(len--){
               int x = q.front()[0];
               int y = q.front()[1];
               int k = q.front()[2];
               q.pop();
               for(int t=0; t<4; ++t){
                   int i = x+dir[t].first;
                   int j = y+dir[t].second;
                   if(i<0||i>=m || j < 0 || j >= n) continue;
                   if(i==m-1 && j==n-1) return step+1;
                   if(grid[i][j]==1){
                       if(k==K)continue;
                       if(visited[i][j][k+1]==1)continue;
                       visited[i][j][k+1]=1;
                       q.push({i, j, k+1});
                   }else{
                       if(visited[i][j][k]==1)continue;
                       visited[i][j][k]=1;
                       q.push({i, j, k});
                   }
               }
           }//end of while
           ++step;
       }
        return -1;
    }
};

File: LC/[1293] Shortest Path in a Grid with Obstacles Elimination.py
from collections import deque


class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        """
        shortest path -> BFS
        Obstacles: additional variable we need to recordld

        # Can we try DP -> if we only go right and down (never go backward) -> might

        """
        visited = set()  # (ii, jj, kk) where ii, jj are the grid coordinates and kk is the # of obstacles used so far
        steps = 0
        dq = deque([(0, 0, 0)])

        M = len(grid)
        N = len(grid[0])
        if M == 1 and N == 1:
            return 0
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        while (sz := len(dq)) > 0:  # take sz for each level
            for _ in range(sz):  # for each level
                # all nodes on same level, level starts from 0. zz is the initial state when etering a new grid
                xx, yy, zz = dq.popleft()
                if xx < 0 or xx >= M or yy < 0 or yy >= N or (xx, yy, zz) in visited:
                    continue
                if grid[xx][yy] == 1:  # if obstacle, consume one k
                    if zz >= k:
                        continue  # if the initial kk state is already k, we cannot proceed from here
                    zz += 1  # consume one k and proceeding
                if xx == M-1 and yy == N-1:  # arrived
                    return steps
                # recording the zz-1 as the initial state on this grid is visited
                visited.add((xx, yy, zz-1))
                for dx, dy in dirs:
                    nx, ny = xx+dx, yy+dy
                    dq.append((nx, ny, zz))

            steps += 1

        return -1
